### ISSUE TITLE
Implement the formulas from the matrix cookbook

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#5858]
See http://orion.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf .  These should all be implemented, e.g., with MatrixExpr.  In particular, currently differentiation with respect to matrices is wrong because it doesn't handle transposes correctly.  For example, compare the output of this to that in the cookbook:

``` python
In [88]: A = MatrixSymbol("A", n, n)

In [89]: x = MatrixSymbol("x", n, 1)

In [90]: diff(A.T*x, x)
Out[90]: A'
```

According to equation 61 in the cookbook, this should just be A.

Original issue for #5858: http://code.google.com/p/sympy/issues/detail?id=2759
Original author: https://code.google.com/u/asmeurer@gmail.com/
Original owner: https://code.google.com/u/asmeurer@gmail.com/

**Comment by asmeurer** [comment]
Also probably related: https://github.com/sympy/sympy/pull/7437. Anyone wanting to work on this should see if that PR should be resurrected.

**Comment by meghana1995** [comment]
@asmeurer . I would like to work on this. It looks like that PR has merge conflicts and some restructuring required. I will look into this

**Comment by Upabjojr** [comment]
https://github.com/sympy/sympy/pull/11570
https://github.com/sympy/sympy/pull/11571

**Comment by jacknlliu** [comment]
Is this process still forward?

**Comment by asmeurer** [comment]
I think this was closed accidentally.

**Comment by Upabjojr** [comment]
Yeah, the implementation is not finished. We need a way to handle the multidimensional expressions. In the cookbook they introduce a `J` matrix of matrices to handle the case `A.diff(A)`.

**Comment by MInner** [comment]
Just wanted to point to www.matrixcalculus.org and the related paper ["Computing Higher Order Derivatives of Matrix and Tensor Expressions"](http://www.matrixcalculus.org/matrixcalculus.pdf). I could not find any traces of an opensource implementation, unfortunately :(

**Comment by Upabjojr** [comment]
The derivative formulae of the cookbook are mostly implemented: https://github.com/sympy/sympy/blob/master/sympy/matrices/expressions/tests/test_derivatives.py
